cpp/arcticdb/storage/azure/azure_mock_client.hpp

```diff
@@ -0,0 +1,60 @@
+/* Copyright 2023 Man Group Operations Limited
+ *
+ * Use of this software is governed by the Business Source License 1.1 included in the file licenses/BSL.txt.
+ *
+ * As of the Change Date specified in that file, in accordance with the Business Source License, use of this software will be governed by the Apache License, version 2.0.
+ */
+
+#pragma once
+
+#include <azure/core/http/curl_transport.hpp>
+#include <azure/core/http/http_status_code.hpp>
+#include <azure/core.hpp>
+#include <azure/storage/blobs.hpp>
+
+#include <arcticdb/storage/storage_utils.hpp>
+#include <arcticdb/storage/azure/azure_client_wrapper.hpp>
+
+namespace arcticdb::storage::azure {
+
+enum class AzureOperation{
+    READ,
+    WRITE,
+    DELETE, // delete_blobs will fail for all objects if one of them triggers a delete failure
+    LIST,
+    EXISTS,
+};
+
+class MockAzureClient : public AzureClientWrapper {
+
+private:
+    // Stores a mapping from blob_name to a Segment.
+    std::map<std::string, Segment> azure_contents;
+public:
+
+    void write_blob(
+            const std::string& blob_name,
+            Segment&& segment,
+            const Azure::Storage::Blobs::UploadBlockBlobFromOptions& upload_option,
+            unsigned int request_timeout) override;
+
+    Segment read_blob(
+            const std::string& blob_name,
+            const Azure::Storage::Blobs::DownloadBlobToOptions& download_option,
+            unsigned int request_timeout) override;
+
+    void delete_blobs(
+            const std::vector<std::string>& blob_names,
+            unsigned int request_timeout) override;
+
+    bool blob_exists(const std::string& blob_name) override;
+
+    Azure::Storage::Blobs::ListBlobsPagedResponse list_blobs(const std::string& prefix) override;
+
+    static std::string get_failure_trigger(
+            const std::string& s3_object_name,
```

Comment:
It would be nice to have a consistent way to trigger failures between `azure_mock_client` and `mock_s3_client`.

What do you think about abstracting out a common `mock_storage_failures.hpp`?
It could contain the following:

1. A common `Operation` enum, which will be a consistent operation enum between `AzureOperation` and `S3Operation` and we should probably use the Azure names because the s3 ones are dumb (sorry, that's my fault).
2. A common parent `MockStorageClient`. So the `MockAzureClient` can inherit both `MockStorageClient` and `AzureClientWrapper`. The `MockStorageClient` can look something like:
```
template<class Key, class Failure>
class MockStorageClient{
private:
     std::map<Key, Segment> contents;
public:
     std::optional<Failure> write(Key key, Segment segment);
     // Other basic operations with the contents (all probably returning something similar to the S3Result to indicate triggered failuer)
    virtual std::optional<Failure> has_failure_trigger(Key key, Operation operation) = 0
}
```
I.e. my thinking is that to build a mock storage what you really need is to just provide a way to build the `Failure`s and have a way to trigger the failures. The other parts of the mock storage should be identical.

So in this case `MockAzureClient : public MockStorageClient<std::string, AzureException>, publick AzureClientWrapper` and would just need to write it's:

- `get_failure_trigger`
- `has_failuer_trigger`
- satisfy the `AzureClientWrapper` interface (which should be trival when using the `MockAzureStorage`'s write, read, delete, etc.)

Let's probably chat offline to figure out the best plan of action.